### PR TITLE
Handle blank error message in response

### PR DESCRIPTION
**Pull request by lukehutch**
This makes `callServerEndpoint` more robust, because I got a 403 response from Serverpod in response to a request, and the Serverpod client code threw an exception because `response.data` was the empty string (`getExceptionFrom` calls `serializationManager.decodeWithType(data)`, which fails for the empty string).

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [ ] I added new tests to check the change I am making.
- [ ] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

_If you have done any breaking changes, make sure to outline them here, so that they can be included in the notes for the next release._

### CODE REVIEW

File: packages/serverpod_client/lib/src/serverpod_client_io.dart [hunk 1]
```diff
@@ -70,11 +69,27 @@ abstract class ServerpodClient extends ServerpodClientShared {
...
+                'Bad request', response.statusCode),
+            HttpStatus.unauthorized => throw ServerpodClientException(
+                'Unauthorized', response.statusCode),
+            HttpStatus.forbidden =>
+              throw ServerpodClientException('Forbidden', response.statusCode),
+            HttpStatus.notFound =>
+              throw ServerpodClientException('Not found', response.statusCode),
+            HttpStatus.internalServerError => throw ServerpodClientException(
+                'Internal server error', response.statusCode),
+            _ => ServerpodClientException('Unknown error', response.statusCode),
```
Comment (Isakdl) [inline packages/serverpod_client/lib/src/serverpod_client_io.dart:84]: Suggestion: Move this code into the method `getExceptionFrom` as it needs to work in web as well and that method is shared between io and the web implementation. 

Then I would also like to see Explicit exception types for each exception here, e.g.

```dart
class ServerpodClientBadRequest extends ServerpodClientException {
  ServerpodClientBadRequest(int statusCode) : super('Bad request', statusCode);
}
```

**Review by Isakdl — CHANGES_REQUESTED** [review]
Thank you for this addition luke! :) I two smaller requests before merging.

File: packages/serverpod_client/lib/src/serverpod_client_exception.dart [hunk 2]
```diff
@@ -14,3 +16,39 @@ class ServerpodClientException implements Exception {
...
+
+/// Thrown if the client created a malformed or invalid request
+/// to the server.
+class ServerpodClientBadRequest extends ServerpodClientException {
+  /// Creates a Bad Request Exception
+  ServerpodClientBadRequest() : super('Bad request', HttpStatus.badRequest);
+}
+
+/// Thrown if the client is not authorized to perform the request.
+
```
Comment (SandPod) [inline packages/serverpod_client/lib/src/serverpod_client_exception.dart:28]: Suggestion(NIT): Remove empty line.

File: packages/serverpod_client/lib/src/serverpod_client_exception.dart [hunk 3]
```diff
@@ -14,3 +16,39 @@ class ServerpodClientException implements Exception {
...
 }
+
+/// Thrown if the client created a malformed or invalid request
+/// to the server.
+class ServerpodClientBadRequest extends ServerpodClientException {
+  /// Creates a Bad Request Exception
+  ServerpodClientBadRequest() : super('Bad request', HttpStatus.badRequest);
+}
+
+/// Thrown if the client is not authorized to perform the request.
```
Comment (SandPod) [inline packages/serverpod_client/lib/src/serverpod_client_exception.dart:27]: Suggestion: Rephrase to `Thrown if the client fails to authenticate and is therefore not authorized to perform the request`.

The current phrasing makes me confuse it with 403: Forbidden.

File: packages/serverpod_client/lib/src/serverpod_client_shared_private.dart [hunk 4]
```diff
@@ -30,6 +32,17 @@ dynamic getExceptionFrom({
   required SerializationManager serializationManager,
   required int statusCode,
 }) {
+  if (data.isEmpty) {
```
Comment (SandPod) [inline packages/serverpod_client/lib/src/serverpod_client_shared_private.dart:35]: Question: Shouldn't this also apply if we fail to decode the passed object?

Basically in both cases below?

Comment (SandPod) [reply packages/serverpod_client/lib/src/serverpod_client_shared_private.dart:35]: Suggestion: add some unit tests for this code that validate this works as expected. Then we could test the case when we fail to decode for example.

**Review by SandPod — CHANGES_REQUESTED** [review]
Nice improvement!

Would love to see some testing around this, we might as well start hardenning this code as well. Also would be nice to try and not drop the exception type if we fail to decode the type.

File: packages/serverpod_client/test/exception_from_test.dart [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+void main() {
+  group('Given a serializable exception when parsing the exception', () {
+    var exception = getExceptionFrom(
+      data:
+          '{"className": "TestException", "data": {"message": "A custom exception"}}',
+      serializationManager: TestSerializationImpl(),
+      statusCode: 400,
+    );
+
+    test('then a base ServerpodClientException is created', () {
```
Comment (SandPod) [inline packages/serverpod_client/test/exception_from_test.dart:48]: chore: Update description so that it matches the test.

Comment (SandPod) [reply packages/serverpod_client/test/exception_from_test.dart:48]: This applies to then next test as well so please check all tests in this file so the descriptions match.

File: packages/serverpod_client/test/exception_from_test.dart [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+      expect(exception, isA<TestException>());
+    });
+
+    test('then the message is an unknown error', () {
+      expect(exception.message, 'A custom exception');
```
Comment (SandPod) [inline packages/serverpod_client/test/exception_from_test.dart:53]: Question: Where is this message defined? Is this part of the framework code?

File: packages/serverpod_client/test/exception_from_test.dart [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    test('then the message is an unknown error', () {
+      expect(exception.message, 'Unknown error, data: malformed data');
+    });
+  });
+
+  group(
+      'Given malformed data with a bad request status code when parsing the exception',
+      () {
+    var exception = getExceptionFrom(
+      data: 'malformed data',
```
Comment (SandPod) [inline packages/serverpod_client/test/exception_from_test.dart:80]: Suggestion: For completeness, please add malformed data request for the different types as well.

Possibly an opportunity for a group.

**Review by SandPod — COMMENTED** [review]
Getting there! Minor comments around the tests and then we are ready to go! 👍

File: packages/serverpod_client/test/exception_from_test.dart (hunk 6 above)
Comment (Isakdl) [reply packages/serverpod_client/test/exception_from_test.dart:53]: It is defined in this test.

**Review by SandPod — APPROVED** [review]
LGTM! 🎉

**Review by Isakdl — APPROVED** [review]
LGTM

**Pull request merged**